datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -77,56 +78,79 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
+      } catch (RetriesExhaustedException ex) {
+        LOG.warn("Attempts to fetch partition throughput timed out");
+        LOG.info("Throughput information unavailable during initial assignment. Falling back to sticky partition assignment");
+        return super.assignPartitions(currentAssignment, datastreamPartitions);
+      }
 
-    // TODO Get unassigned partitions
-    // Calculating unassigned partitions
-    List<String> unassignedPartitions = new ArrayList<>();
+      // Task count update happens only on initial assignment (when datastream makes the STOPPED -> READY transition).
+      // The calculation is based on the maximum of:
+      //   (1) Tasks already allocated for the datastream
+      //   (2) Partition number based estimate, if the appropriate config is enabled
+      //   (3) Throughput based task count estimate
+      int numTasksNeeded = taskCount;
+      if (_enablePartitionNumBasedTaskCountEstimation) {
+        numTasksNeeded = getTaskCountEstimateBasedOnNumPartitions(datastreamPartitions, taskCount);
+      }
+
+      LoadBasedTaskCountEstimator estimator = new LoadBasedTaskCountEstimator(_taskCapacityMBps, _taskCapacityUtilizationPct);
+      numTasksNeeded = Math.max(numTasksNeeded, estimator.getTaskCount(clusterThroughputInfo, assignedPartitions,
+          unassignedPartitions));
+
+      // Task count is validated against max tasks config
+      numTasksNeeded = validateNumTasksAgainstMaxTasks(datastreamPartitions, numTasksNeeded);
+      if (numTasksNeeded > taskCount) {
+        updateNumTasksAndForceTaskCreation(datastreamPartitions, numTasksNeeded, taskCount);
+      }
+    }
 
+    // TODO Implement metrics
     // Doing assignment
     LoadBasedPartitionAssigner partitionAssigner = new LoadBasedPartitionAssigner();
-    return partitionAssigner.assignPartitions(clusterThroughputInfo, currentAssignment,
-        unassignedPartitions, datastreamPartitions);
+    Map<String, Set<DatastreamTask>> newAssignment = partitionAssigner.assignPartitions(clusterThroughputInfo,
+        currentAssignment, unassignedPartitions, datastreamPartitions);
+    LOG.info("new assignment info, assignment: {}", newAssignment);
+    partitionSanityChecks(newAssignment, datastreamPartitions);
+    return newAssignment;
   }
 
-  private Map<String, ClusterThroughputInfo> fetchPartitionThroughputInfo() {
+  private ClusterThroughputInfo fetchPartitionThroughputInfo(DatastreamGroup datastreamGroup) {
+    AtomicInteger attemptNum = new AtomicInteger(0);
     return PollUtils.poll(() -> {
       try {
-        return _throughputProvider.getThroughputInfo();
+        return _throughputProvider.getThroughputInfo(datastreamGroup);
       } catch (Exception ex) {
-        // TODO print exception and retry count
-        LOG.warn("Failed to fetch partition throughput info.");
+        attemptNum.set(attemptNum.get() + 1);
+        LOG.warn("Failed to fetch partition throughput info on attempt {}.", attemptNum.get(), ex);
```

Comment:
This LOG.warn call is going to `public void warn(String format, Object arg1, Object arg2);` instead of `public void warn(String msg, Throwable t);`. This warn may not print.